DataFormats/Common/interface/RefToBase.h

```diff
@@ -228,43 +233,20 @@ namespace edm {
     return holder_->key();
   }
 
-  namespace {
-    // If the template parameters are classes or structs they should be
-    // related by inheritance, otherwise they should be the same type.
-    template <typename T, typename U>
-    typename std::enable_if<std::is_class<T>::value>::type checkTypeCompatibility() {
-      static_assert(std::is_base_of<T, U>::value || std::is_base_of<U, T>::value,
-                    "RefToBase::castTo error element types are not related by inheritance");
-    }
-
-    template <typename T, typename U>
-    typename std::enable_if<!std::is_class<T>::value>::type checkTypeCompatibility() {
-      static_assert(std::is_same<T, U>::value, "RefToBase::castTo error non-class element types are not the same");
-    }
-
-    // Convert the pointer types, use dynamic_cast if they are classes
-    template <typename T, typename OUT>
-    typename std::enable_if<std::is_class<T>::value, OUT const*>::type convertTo(T const* t) {
-      return dynamic_cast<OUT const*>(t);
-    }
-
-    template <typename T, typename OUT>
-    typename std::enable_if<!std::is_class<T>::value, OUT const*>::type convertTo(T const* t) {
-      return t;
```

Comment:
(this may be just to note myself) Earlier there was a case if `T` is non-class type, the `T const*` is directly converted to `OUT const*`. For non-class types I'd expect this to fail to compile if `T != OUT`. So in the new code the `std::is_same_v<T, OUT>` is equivalent.

On the other hand, in https://github.com/cms-sw/cmssw/issues/42737 it was found that we don't have cases where `T` would not be a class type.

Reply:
The call to `castTo` also calls `checkTypeCompatiblity<>` which requires not class types to have identical types. So the combined calls to `chekTypeCompatibility` and `convertTo` require the following
- for classes
   - must be related by inheritance
- for non-classes
   - types must be the same

turns out for non classes, calls to `std::is_base_of` always return `false` and for classes, passing the same class to `std::is_base_of` return `true`. So the new constraint on `castTo` should be identical to the combined constraints from the old code.